### PR TITLE
priorbox_8 ngraph and inference-engine parts

### PR DESCRIPTION
**Pull request by liubo-intel**
### Details:
 - *New version of PriorBox operation to add new 'min_max_aspect_ratios_order' attribute*
 - *Spec : https://github.com/openvinotoolkit/openvino/pull/8000*

### Tickets:
 - *66645*
 - *66665*
 - *66666*

### CODE REVIEW

**Comment by liubo-intel** [comment]
Hi，@  : could you please help have a look at this PR for priorbox_8? as you can find that priorbox operation will fallback to runtime::reference implement for both the previous version and this new version, so in this PR, I also modified "plugin/cpu/shared_tests_instances" of it to test cpu plugin of this operation new version. If you recommand other more reviewers, please also help add them in this PR, thanks.

File: ngraph/core/include/openvino/op/prior_box.hpp [hunk 1]
```diff
@@ -66,5 +66,66 @@ class OPENVINO_API PriorBox : public Op {
...
+
+    static int64_t number_of_priors(const Attributes& attrs);
+
+    static std::vector<float> normalized_aspect_ratio(const std::vector<float>& aspect_ratio, bool flip);
+    const Attributes& get_attrs() const {
+        return m_attrs;
+    }
+    bool visit_attributes(AttributeVisitor& visitor) override;
+    OPENVINO_SUPPRESS_DEPRECATED_START
+    bool evaluate(const HostTensorVector& outputs, const HostTensorVector& inputs) const override;
```
[the remark this answers is not in the document]
Comment (ilyachur) [reply ngraph/core/include/openvino/op/prior_box.hpp:122]: Hi, In the PR https://github.com/openvinotoolkit/openvino/pull/8297 I enabled a support of new evaluate methods in the CPU plugin.
Could you try to use new API?

[the remark this answers is not in the document]
Comment (liubo-intel) [reply ngraph/core/include/openvino/op/prior_box.hpp]: updated in commit 68eb0e8

File: ngraph/core/include/openvino/op/prior_box.hpp [hunk 2]
```diff
@@ -58,13 +58,74 @@ class OPENVINO_API PriorBox : public Op {
     }
     bool visit_attributes(AttributeVisitor& visitor) override;
     OPENVINO_SUPPRESS_DEPRECATED_START
-    bool evaluate(const HostTensorVector& outputs, const HostTensorVector& inputs) const override;
+    bool evaluate(ov::runtime::TensorVector& outputs, const ov::runtime::TensorVector& inputs) const override;
```
Comment (ilyachur) [inline ngraph/core/include/openvino/op/prior_box.hpp:61]: Please don't change the v0operation. It is not a backward compatible change.

File: ngraph/core/include/openvino/op/prior_box.hpp [hunk 3]
```diff
@@ -58,13 +58,74 @@ class OPENVINO_API PriorBox : public Op {
...
+    std::shared_ptr<Node> clone_with_new_inputs(const OutputVector& new_args) const override;
+
+    static int64_t number_of_priors(const Attributes& attrs);
+
+    static std::vector<float> normalized_aspect_ratio(const std::vector<float>& aspect_ratio, bool flip);
+    const Attributes& get_attrs() const {
+        return m_attrs;
+    }
+    bool visit_attributes(AttributeVisitor& visitor) override;
+    OPENVINO_SUPPRESS_DEPRECATED_START
```
Comment (ilyachur) [inline ngraph/core/include/openvino/op/prior_box.hpp:121]: Please remove suppress macros

File: ngraph/core/reference/include/ngraph/runtime/reference/prior_box.hpp [hunk 4]
```diff
@@ -22,8 +22,71 @@ static inline float clip_less(float x, float threshold) {
...
+    return num_priors;
+}
+
 template <typename T>
-void prior_box(const T* data, const T* img, float* dst_data, const Shape& out_shape, const op::PriorBoxAttrs& attrs) {
+void prior_box(const T* data,
+               const T* img,
+               float* dst_data,
+               const Shape& out_shape,
+               const std::vector<float>& attrs_min_size,
```
Comment (ilyachur) [inline ngraph/core/reference/include/ngraph/runtime/reference/prior_box.hpp:77]: Why don't pass the Attribute structure?

File: ngraph/core/src/node.cpp [hunk 5]
```diff
@@ -782,10 +820,16 @@ bool ov::Node::constant_fold(OutputVector& output_values, const OutputVector& in
         auto tensor = make_shared<HostTensor>(output.get_element_type(), output.get_partial_shape());
         output_tensors.push_back(tensor);
     }
+
+    ov::runtime::TensorVector input_runtime_tensors = create_tmp_tensors(input_tensors);
```
Comment (ilyachur) [inline ngraph/core/src/node.cpp:824]: You don't need create temporary buffers. Please create `runtime::TesorVector` instead of `ngraph::HostTensorVector` and call new Evaluate here.

File: ngraph/core/src/node.cpp [hunk 6]
```diff
@@ -782,10 +820,16 @@ bool ov::Node::constant_fold(OutputVector& output_values, const OutputVector& in
...
+    ov::runtime::TensorVector output_runtime_tensors = create_tmp_tensors(output_tensors);
+
     OPENVINO_SUPPRESS_DEPRECATED_START
```
Comment (ilyachur) [inline ngraph/core/src/node.cpp:827]: Please remove suppress macro if you use new evaluate.

File: ngraph/core/src/node.cpp [hunk 7]
```diff
@@ -770,26 +770,32 @@ bool ov::Node::constant_fold(OutputVector& output_values, const OutputVector& in
...
+        }
     }
-    HostTensorVector output_tensors;
-    OutputVector output_constants;
+
+    ov::runtime::TensorVector output_tensors;
     for (const auto& output : outputs()) {
-        auto tensor = make_shared<HostTensor>(output.get_element_type(), output.get_partial_shape());
-        output_tensors.push_back(tensor);
+        auto host_tensor = ov::runtime::Tensor(output.get_element_type(), output.get_shape());
```
Comment (ilyachur) [inline ngraph/core/src/node.cpp:786]: It is not correct, because output shape can be dynamic.
Please add a code for dynamic shapes.
Example:
https://github.com/openvinotoolkit/openvino/blob/a115f2f4ea5cdee224e5fb0fac2f2a459fa61455/ngraph/core/src/function.cpp#L436

Comment (liubo-intel) [reply ngraph/core/src/node.cpp]: added in commit 94306e1

**Comment by liubo-intel** [comment]
Hi, @ilyachur  : As we synced offline, using new 'runtime::Tensor' API for 'PriorBox8' looks like bring in other Ops' test case failure, so I reverted to use 'HostTensor' for this PR, and will use another ticket to track using New API issue.

File: ngraph/test/runtime/interpreter/evaluates_map.cpp [hunk 8]
```diff
@@ -1529,6 +1529,33 @@ bool evaluate(const shared_ptr<op::v0::Elu>& op, const HostTensorVector& outputs
 template <element::Type_t ET>
 bool evaluate(const shared_ptr<op::v0::PriorBox>& op, const HostTensorVector& outputs, const HostTensorVector& inputs) {
     using T = typename element_type_traits<ET>::value_type;
+    op::v0::PriorBox::Attributes attrs = op->get_attrs();
```
Comment (ilyachur) [inline ngraph/test/runtime/interpreter/evaluates_map.cpp:1532]: Looks like you don't need tp have these methods in evaluates map because both priorbox ops have evaluate methods

File: runtime/bindings/python/src/openvino/opset8/ops.py [hunk 9]
```diff
@@ -414,3 +414,108 @@ def gather_nd(
     }
 
     return _get_node_factory_opset8().create("GatherND", inputs, attributes)
+
+
+@nameable_op
+def prior_box(
```
Comment (ilyachur) [inline runtime/bindings/python/src/openvino/opset8/ops.py:420]: We also need to add this op to legacy python API

**Review by ilyachur — COMMENTED** [review]
@GlebKazantaev Please take a look to transformation part

File: ngraph/test/runtime/interpreter/evaluates_map.cpp (hunk 8 above)
Comment (liubo-intel) [reply ngraph/test/runtime/interpreter/evaluates_map.cpp]: removed these evaluates in commit abb47a7

File: runtime/bindings/python/src/openvino/opset8/ops.py (hunk 9 above)
Comment (liubo-intel) [reply runtime/bindings/python/src/openvino/opset8/ops.py:420]: added in abb47a7

File: inference-engine/src/transformations/src/transformations/common_optimizations/common_optimizations.cpp [hunk 10]
```diff
@@ -147,6 +148,7 @@ bool ngraph::pass::CommonOptimizations::run_on_function(std::shared_ptr<ngraph::
     decomp->add_matcher<ngraph::pass::DropoutWithRandomUniformReplacer>();
     decomp->set_name("ngraph::pass::CommonDecompositions");
 
+    manager.register_pass<ngraph::pass::ConvertPriorBox8To0>();  // not plugins implemented priorbox8
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/common_optimizations.cpp:151]: Please move it to the section below where other downgrade passes are located.

File: inference-engine/src/transformations/src/transformations/op_conversions/convert_prior_box_v8_to_v0.cpp [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+        attrs_v0.clip = attrs_v8.clip;
+        attrs_v0.flip = attrs_v8.flip;
+        attrs_v0.step = attrs_v8.step;
+        attrs_v0.offset = attrs_v8.offset;
+        attrs_v0.variance = attrs_v8.variance;
+        attrs_v0.scale_all_sizes = attrs_v8.scale_all_sizes;
+
+        auto output_size = prior_box_v8_node->input_value(0);
+        auto image_size = prior_box_v8_node->input_value(1);
+        auto prior_box_v0 = std::make_shared<ngraph::opset1::PriorBox>(output_size, image_size, attrs_v0);
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/op_conversions/convert_prior_box_v8_to_v0.cpp:46]:
```suggestion
        auto prior_box_v0 = std::make_shared<ngraph::opset1::PriorBox>(prior_box_v8_node->input_value(0), prior_box_v8_node->input_value(1), attrs_v0);
```

File: inference-engine/src/transformations/src/transformations/common_optimizations/common_optimizations.cpp (hunk 10 above)
Comment (liubo-intel) [reply inference-engine/src/transformations/src/transformations/common_optimizations/common_optimizations.cpp]: updated in commit 5cf6467

File: inference-engine/src/transformations/src/transformations/op_conversions/convert_prior_box_v8_to_v0.cpp (hunk 11 above)
Comment (liubo-intel) [reply inference-engine/src/transformations/src/transformations/op_conversions/convert_prior_box_v8_to_v0.cpp]: updated in commit 5cf6467

File: runtime/bindings/python/src/compatibility/ngraph/opset8/ops.py [hunk 12]
```diff
@@ -417,3 +417,108 @@ def gather_nd(
...
+        ("step", False, np.floating, is_non_negative_value),
+        ("variance", False, np.floating, is_positive_value),
+        ("scale_all_sizes", False, np.bool_, None),
+        ("fixed_ratio", False, np.floating, is_positive_value),
+        ("fixed_size", False, np.floating, is_positive_value),
+        ("density", False, np.floating, is_positive_value),
+        ("min_max_aspect_ratios_order", False, np.bool_, None),
+    ]
+
+    check_valid_attributes("PriorBox", attrs, requirements)
```
Comment (ilyachur) [inline runtime/bindings/python/src/compatibility/ngraph/opset8/ops.py:522]: I think you also need to update init file for opset8
https://github.com/openvinotoolkit/openvino/blob/master/runtime/bindings/python/src/compatibility/ngraph/opset8/__init__.py

Comment (liubo-intel) [reply runtime/bindings/python/src/compatibility/ngraph/opset8/ops.py:522]: updated in commit 6f64144

**Review by GlebKazantaev — APPROVED** [review]
Transformation part looks good.

**Review by ilyachur — APPROVED** [review]

**Review by iefode — APPROVED** [review]

File: ngraph/core/include/ngraph/op/prior_box.hpp [hunk 13]
```diff
@@ -9,10 +9,11 @@
 
 namespace ngraph {
 namespace op {
-using PriorBoxAttrs = ov::op::v0::PriorBox::Attributes;
```
Comment (ilyachur) [inline ngraph/core/include/ngraph/op/prior_box.hpp:12]: Please revert this change. IIt should allow to fix opencv

File: ngraph/core/include/ngraph/op/prior_box.hpp [hunk 14]
```diff
@@ -9,10 +9,11 @@
...
 namespace v0 {
 using ov::op::v0::PriorBox;
 }  // namespace v0
-using v0::PriorBox;
```
Comment (ilyachur) [inline ngraph/core/include/ngraph/op/prior_box.hpp:16]: Please don't remove this using

File: ngraph/core/include/ngraph/op/prior_box.hpp (hunk 13 above)
Comment (liubo-intel) [reply ngraph/core/include/ngraph/op/prior_box.hpp]: updated in commit 7b99a21

**Review by jane-intel — APPROVED** [review]
Please explicitly confirm that this doesn't brake TEMPLATE plugin inference of models with PriorBox-1 -- and also PriorBox-8 should work seamlessly on TEMPLATE plugin too.

**Comment by liubo-intel** [comment]
> Please explicitly confirm that this doesn't brake TEMPLATE plugin inference of models with PriorBox-1 -- and also PriorBox-8 should work seamlessly on TEMPLATE plugin too.

Hi, @jane-intel : we have added 'templateFuncTests' for both ['PriorBox-1' and 'PriorBox-8' ](https://github.com/openvinotoolkit/openvino/pull/8114/files#diff-a4d2072da718d83bb410fc8d6913d90225fccf5ce3022561fca27bfea980784b), so I think both PriorBox-1  and  PriorBox-8 should work seamlessly on TEMPLATE plugin.

**Review by ilyachur — COMMENTED** [review]
Why did you update onednn_gpu? Please revert these changes

**Comment by liubo-intel** [comment]
> Why did you update onednn_gpu? Please revert these changes

Hi, Ilya:  I'm not so sure why this 'thirdparty' git changes happened, maybe I forgot update --init before push commit... Have revert this changes in commit 3ad71d3.

File: src/core/include/openvino/op/prior_box.hpp [hunk 15]
```diff
@@ -66,5 +66,66 @@ class OPENVINO_API PriorBox : public Op {
...
 };
 }  // namespace v0
+
+namespace v8 {
+/// \brief Layer which generates prior boxes of specified sizes
+/// normalized to input image size
+class OPENVINO_API PriorBox : public Op {
+public:
+    OPENVINO_OP("PriorBox", "opset8");
+    BWDCMP_RTTI_DECLARATION;
```
Comment (ilyachur) [inline src/core/include/openvino/op/prior_box.hpp:76]: We don't need this `BWDCMP_RTTI_DECLARATION` macro

**Review by ilyachur — APPROVED** [review]

File: src/core/include/openvino/op/prior_box.hpp (hunk 15 above)
Comment (liubo-intel) [reply src/core/include/openvino/op/prior_box.hpp:76]: will  remove this macro in  next PR
